lib/client/api.go

```diff
@@ -3581,6 +3589,71 @@ func (tc *TeleportClient) Login(ctx context.Context) (*Key, error) {
 	return key, nil
 }
 
+// AttemptDeviceLogin attempts device authentication for the current device.
+// It expects to receive the latest activated [key], as acquired via
+// [TeleportClient.Login], and augments the certificates within the key with
+// device extensions.
+//
+// If successful, the new device certificates are automatically activated (using
+// [TeleportClient.ActivateKey].)
+//
+// A nil response from this method doesn't mean that device authentication was
+// successful, as skipping the ceremony is valid for various reasons (Teleport
+// cluster doesn't support device authn, device wasn't enrolled, etc).
+// Use [TeleportClient.DeviceLogin] if you want more control over process.
+func (tc *TeleportClient) AttemptDeviceLogin(ctx context.Context, key *Key) error {
+	newCerts, err := tc.DeviceLogin(ctx, &devicepb.UserCertificates{
+		// Augment the SSH certificate.
+		// The TLS certificate is already part of the connection.
+		SshAuthorizedKey: key.Cert,
+	})
+	if err != nil {
+		log.WithError(err).Debug("Device Trust: device authentication failed")
+		return nil // Swallowed on purpose.
+	}
+
+	log.Debug("Device Trust: acquired augmented user certificates")
+	cp := *key
+	cp.Cert = newCerts.SshAuthorizedKey
+	cp.TLSCert = pem.EncodeToMemory(&pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: newCerts.X509Der,
+	})
+	return trace.Wrap(tc.ActivateKey(ctx, &cp))
+}
+
+// dtAuthnRunCeremony is used to fake device authentication for tests.
+var dtAuthnRunCeremony = dtauthn.RunCeremony
+
+// DeviceLogin attempts to authenticate the current device with Teleport.
+// The device must be previously registered and enrolled for the authentication
+// to succeed (see `tsh device enroll`).
+//
+// DeviceLogin may fail for a variety of reasons, some of them legitimate
+// (non-Enterprise cluster, Device Trust is disabled, etc). Because of that, a
+// failure in this method may not warrant failing a broader action (for example,
+// `tsh login`).
+//
+// Device Trust is a Teleport Enterprise feature.
+func (tc *TeleportClient) DeviceLogin(ctx context.Context, certs *devicepb.UserCertificates) (*devicepb.UserCertificates, error) {
+	// TODO(codingllama): Determine if Device Trust is supported/enabled.
+	//  One should only pay for the roundtrip if they actually use the feature.
```

Comment:
I thinnk the general agreement so far is that it's ok for the login process to do extra roundtrips; it would be nice to avoid the extra connection to the proxy tho.

Reply:
Yep, my current plan is to plug it into Ping, so we avoid all additional work under some settings. I wanted to get it into working shape first, though.